components/authz-service/storage/v2/postgres/postgres.go

```diff
@@ -950,6 +950,211 @@ func (p *pg) insertRoleWithQuerier(ctx context.Context, role *v2.Role, q Querier
 	return nil
 }
 
+/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */
+/* * * * * * * * * * * * * * * * * *    Rules    * * * * * * * * * * * * * * * * * * * */
+/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */
+
+func (p *pg) CreateRule(ctx context.Context, rule *v2.Rule) (*v2.Rule, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	tx, err := p.db.BeginTx(ctx, nil)
+	if err != nil {
+		return nil, p.processError(err)
+	}
+
+	row := tx.QueryRowContext(ctx,
+		`INSERT INTO iam_project_rules (id, project_id, name, type) VALUES ($1, $2, $3, $4) RETURNING db_id;`,
+		rule.ID, rule.ProjectID, rule.Name, rule.Type.String())
+	var ruleDbID string
+	if err := row.Scan(&ruleDbID); err != nil {
+		return nil, p.processError(err)
+	}
+
+	for _, condition := range rule.Conditions {
+		_, err := tx.ExecContext(ctx,
+			`INSERT INTO iam_rule_conditions (rule_id, value, attribute, operator) VALUES ($1, $2, $3, $4);`,
+			ruleDbID, condition.Value, condition.Attribute.String(), condition.Operator.String(),
+		)
+		if err != nil {
+			return nil, p.processError(err)
+		}
+	}
+
+	err = tx.Commit()
+	if err != nil {
+		return nil, storage_errors.NewErrTxCommit(err)
+	}
+
+	// Currently, we don't change anything from what is passed in.
+	return rule, nil
+}
+
+func (p *pg) UpdateRule(ctx context.Context, rule *v2.Rule) (*v2.Rule, error) {
+	projectsFilter, err := projectsListFromContext(ctx)
+	if err != nil {
+		return nil, p.processError(err)
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	tx, err := p.db.BeginTx(ctx, nil)
+	if err != nil {
+		return nil, p.processError(err)
+	}
+
+	row := tx.QueryRowContext(ctx,
+		`UPDATE iam_project_rules SET (name, type) = ($1, $2)
+			WHERE id = $3 AND projects_match_for_rule(project_id, $4) RETURNING db_id, project_id`,
+		rule.Name, rule.Type.String(), rule.ID, pq.Array(projectsFilter))
+	var ruleDbID string
+	var projectID string
+	if err := row.Scan(&ruleDbID, &projectID); err != nil {
+		return nil, p.processError(err)
+	}
+
+	// If they tried to change the project_id, abort the transaction with an error.
+	if projectID != rule.ProjectID {
```

Comment:
The order of operations feels awkward, here, but I see how you've added this flow to simplify fetching/validating the project_id. Can't think of a better way, though. 🤷‍♂

Reply:
It's just saving us a fetch query to get it back out of the delete, should be faster this way.